stan/math/rev/functor/cvodes_integrator.hpp

```diff
@@ -23,50 +22,159 @@ namespace math {
  * methods).
  *
  * @tparam Lmm ID of ODE solver (1: ADAMS, 2: BDF)
+ * @tparam F Type of ODE right hand side
+ * @tparam T_y0 Type of scalars for initial state
+ * @tparam T_param Type of scalars for parameters
+ * @tparam T_t0 Type of scalar of initial time point
+ * @tparam T_ts Type of time-points where ODE solution is returned
  */
-template <int Lmm>
+template <int Lmm, typename F, typename T_y0, typename T_t0, typename T_ts,
+          typename... T_Args>
 class cvodes_integrator {
- public:
-  cvodes_integrator() {}
+  using T_Return = return_type_t<T_y0, T_t0, T_ts, T_Args...>;
+  using T_y0_t0 = return_type_t<T_y0, T_t0>;
+
+  const char* function_name_;
+  const F& f_;
+  const Eigen::Matrix<T_y0_t0, Eigen::Dynamic, 1> y0_;
+  const T_t0 t0_;
+  const std::vector<T_ts>& ts_;
+  std::tuple<const T_Args&...> args_tuple_;
+  std::tuple<plain_type_t<decltype(value_of(std::declval<T_Args>()))>...>
+      value_of_args_tuple_;
+  const size_t N_;
+  std::ostream* msgs_;
+  double relative_tolerance_;
+  double absolute_tolerance_;
+  long int max_num_steps_;  // NOLINT(runtime/int)
+
+  const size_t num_y0_vars_;
+  const size_t num_args_vars_;
+
+  coupled_ode_system<F, T_y0_t0, T_Args...> coupled_ode_;
+
+  std::vector<double> coupled_state_;
+  N_Vector nv_state_;
+  N_Vector* nv_state_sens_;
+  SUNMatrix A_;
+  SUNLinearSolver LS_;
 
   /**
-   * Return the solutions for the specified system of ordinary
-   * differential equations given the specified initial state,
-   * initial times, times of desired solution, and parameters and
-   * data, writing error and warning messages to the specified
-   * stream.
-   *
-   * This function is templated to allow the initials to be
-   * either data or autodiff variables and the parameters to be data
-   * or autodiff variables.  The autodiff-based implementation for
-   * reverse-mode are defined in namespace <code>stan::math</code>
-   * and may be invoked via argument-dependent lookup by including
-   * their headers.
-   *
-   * The solver used is based on the backward differentiation
-   * formula which is an implicit numerical integration scheme
-   * appropriate for stiff ODE systems.
-   *
-   * @tparam F type of ODE system function.
-   * @tparam T_initial type of scalars for initial values.
-   * @tparam T_param type of scalars for parameters.
-   * @tparam T_t0 type of scalar of initial time point.
-   * @tparam T_ts type of time-points where ODE solution is returned.
+   * Implements the function of type CVRhsFn which is the user-defined
+   * ODE RHS passed to CVODES.
+   */
+  static int cv_rhs(realtype t, N_Vector y, N_Vector ydot, void* user_data) {
+    cvodes_integrator* integrator = static_cast<cvodes_integrator*>(user_data);
+    integrator->rhs(t, NV_DATA_S(y), NV_DATA_S(ydot));
+    return 0;
+  }
```

Comment:
`void*`? Is that really voided out / raw memory? I'm now following how that voided `user_data` then becomes a `cvodes_integrator*`

Reply:
Is `NV_DATA_S(y)` etc a C or C++ type (aka do we have to worry about destruction here)?

Reply:
These callbacks are how we provide ODE right hand side evaluations/Jacobians/etc. to CVODES. See [here](https://github.com/stan-dev/math/blob/46465932d90afbd63bc7c7a9c9e49758515504ef/stan/math/rev/functor/cvodes_integrator.hpp#L273) for instance.

I think the NV_DATA_S stuff is fine. It's like C stuff: http://runge.math.smu.edu/arkode_dev/doc/guide/build/html/nvectors/NVector_Serial.html

Reply:
Okay I was thinking about #1950 but as long as those are destroyed during RAII that's fine